File: serialize-and-deserialize-bst.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Codec:

    def serialize(self, root):
        vals = []

        def preOrder(node):
            if node:
                vals.append(node.val)
                preOrder(node.left)
                preOrder(node.right)

        preOrder(root)

        return ' '.join(map(str, vals))

    # O( N ) since each val run build once
    def deserialize(self, data):
        vals = collections.deque(int(val) for val in data.split())

        def build(minVal, maxVal):
            if vals and minVal < vals[0] < maxVal:
                val = vals.popleft()
                node = TreeNode(val)
                node.left = build(minVal, val)
                node.right = build(val, maxVal)
                return node

        return build(float('-infinity'), float('infinity'))

    # O( N*log(N) )
    # def deserialize(self, data):
    #     if data == '':
    #         return None

    #     def insert(node, val):
    #         if val < node.val:
    #             if node.left is None:
    #                 node.left = TreeNode(val)
    #             else:
    #                 insert(node.left, val)
    #         else:
    #             if node.right is None:
    #                 node.right = TreeNode(val)
    #             else:
    #                 insert(node.right, val)

    #     vals = [int(val) for val in data.split(' ')]

    #     root = TreeNode(vals[0])

    #     for val in vals[1:]:
    #         insert(root, val)

    #     return root


# Your Codec object will be instantiated and called as such:
# codec = Codec()
# codec.deserialize(codec.serialize(root))

File: serialize-and-deserialize-bst.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */

/**
 * Encodes a tree to a single string.
 *
 * @param {TreeNode} root
 * @return {string}
 */
var serialize = function(root) {
  const values = []
  function preOrder (node) {
    if (!node) {
      return
    }
    values.push(node.val.toString())
    preOrder(node.left)
    preOrder(node.right)
  }
  preOrder(root)
  return values.join(',')
};

/**
 * Decodes your encoded data to tree.
 *
 * @param {string} data
 * @return {TreeNode}
 */
var deserialize = function(data) {
  if (!data) {
    return null
  }

  function insert (node, val) {
    if (!node) {
      return new TreeNode(val)
    }
    if (val > node.val) {
      node.right = insert(node.right, val)
    } else {
      node.left = insert(node.left, val)
    }
    return node
  }
  let root = null
  data.split(',').forEach((str) => {
    root = insert(root, parseFloat(str))
  })
  return root
};

/**
 * Your functions will be called as such:
 * deserialize(serialize(root));
 */
